aten/src/ATen/core/dispatch/Dispatcher.cpp

```diff
@@ -498,24 +498,44 @@ std::vector<OperatorName> Dispatcher::getRegistrationsForDispatchKey(c10::option
   });
 }
 
-int64_t Dispatcher::sequenceNumberForRunningRecordFunction(DispatchKey dispatchKey) {
+int64_t Dispatcher::sequenceNumberForRunningRecordFunction(DispatchKey dispatchKey, DispatchKeySet dispatchKeySet) {
   int64_t seq_num = -1;
   // Setting sequence number in the Autograd case to associate
   // the forward range with the corresponding Autograd's node
-  if (isIncludedInAlias(dispatchKey, DispatchKey::Autograd) && at::GradMode::is_enabled()) {
+
+  // the usual case, where we're dealing with Autograd dispatchKey
+  bool dispatchHasAutograd = isIncludedInAlias(dispatchKey, DispatchKey::Autograd);
+
+  // Note [Sequence Numbers for Python Subclasses]
+  // See https://github.com/pytorch/pytorch/issues/121758 for more details
+  // The first entry into the dispatch is usually going to be a call(), but
+  // after that we usually redispatch().
+  // We only profile call() and callBoxed(), not redispatch.*()
+  // dispatcher calls, and when running with python subclasses, the top-level
+  // call is a PythonTLSSnapshot call, not an Autograd call. That means that
+  // we need special handling for subclasses if we want to collect the
+  // sequence number for operators on python subclasses.
+  //
+  // This is the special case: if we're profiling a PythonTLSSnapshot call
+  // and we have an autograd dispatch key in the dispatchKeySet, then we'll
```

Comment:
I'm curious why the logic here is not simplified to always get the seq_num unless autograd is in the exclude set?

Reply:
@albanD that's a good point.

My goal was to only capture the sequence_nr once during the forward pass in profiler events. My guess was that this is a safe assumption for PythonTLSSnapshot, but that it might not be true in general (i.e. if we `call` PythonTLSSnapshot, then we won't later `call` again on or above Autograd).

But I guess maybe I should instead just always get the sequence_nr and then fix the profiler logic so that it handles duplicate sequence_nr in forward. Let me experiment to see if I can repro this case and try to fix it

Reply:
From what I recall, the profiler already has to handle duplicates since the current logic can actually trigger multiple peek() of the same number when CompositeImplicit ops are involved.